cpp/src/io/comp/uncomp.cpp

```diff
@@ -579,16 +579,14 @@ class HostDecompressor_SNAPPY : public HostDecompressor {
  */
 /* ----------------------------------------------------------------------------*/
 
-HostDecompressor *HostDecompressor::Create(int stream_type)
+std::unique_ptr<HostDecompressor> HostDecompressor::Create(int stream_type)
 {
-  HostDecompressor *decompressor;
   switch (stream_type) {
-    case IO_UNCOMP_STREAM_TYPE_GZIP: decompressor = new HostDecompressor_ZLIB(true); break;
-    case IO_UNCOMP_STREAM_TYPE_INFLATE: decompressor = new HostDecompressor_ZLIB(false); break;
-    case IO_UNCOMP_STREAM_TYPE_SNAPPY: decompressor = new HostDecompressor_SNAPPY(); break;
-    default: decompressor = nullptr; break;
+    case IO_UNCOMP_STREAM_TYPE_GZIP: return std::make_unique<HostDecompressor_ZLIB>(true);
+    case IO_UNCOMP_STREAM_TYPE_INFLATE: return std::make_unique<HostDecompressor_ZLIB>(false);
+    case IO_UNCOMP_STREAM_TYPE_SNAPPY: return std::make_unique<HostDecompressor_SNAPPY>();
   }
-  return decompressor;
+  return {};
```

Comment:
Should you `throw` / `CUDF_FAIL` here?

Reply:
This keeps the original behavior - return null if the compression type is not supported.

I'll check if null is actually signalling an error here.

Reply:
Yup, returning null here causes an error later on, when trying to parse the data that was never successfully read. Added `CUDF_FAIL`, removed code that assumes that the decompressor can be null.